File: linked_list/linked_list.py
class Node:
    def __init__(self, v):
        self.value = v
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def add_in_tail(self, item):
        if self.head is None:
            self.head = item
        else:
            self.tail.next = item
        self.tail = item

    def print_all_nodes(self):
        node = self.head
        while node != None:
            print(node.value)
            node = node.next

    def find(self, val):
        node = self.head
        while node is not None:
            if node.value == val:
                return node
            node = node.next
        return None

    def find_all(self, val):
        res = []
        node = self.head
        while node is not None:
            if node.value == val:
                res.append(node)
            node = node.next
        return res

    def delete(self, val, all=False):
        if self.head is None:
            return None

        while self.head.value == val:
            self.head = self.head.next
            if self.head is None:
                self.tail = None
                return
            if not all:
                return

        prev = self.head
        node = self.head
        while node is not None:
            if node.value == val:
                if all:
                    prev.next = node.next if node.next is not None and node.next.value != val else prev.next
                    node = node.next
                    if node is None:
                        prev.next = None
                        self.tail = prev
                    continue
                else:
                    prev.next = node.next
                if prev.next is None:
                    self.tail = prev
                if not all:
                    return

            prev = node 
            node = node.next
        return None

    def clean(self):
        self.head = None
        self.tail = None

    def len(self):
        res = 0
        node = self.head
        while node is not None:
            res += 1
            node = node.next
        return res

    def insert(self, afterNode, newNode):
        if self.tail is None:
            self.tail = newNode
        if afterNode is None:
            buff = self.head
            self.head = newNode
            self.head.next = buff
            return
        node = self.head
        while node is not None:
            if node == afterNode:
                buff = node.next
                node.next = newNode
                newNode.next = buff
                if newNode.next is None:
                    self.tail = newNode
                return
            node = node.next

File: asd_go/linkedlist/linked_list.go
package main

import (
	"errors"
	"os"
	"reflect"
)

type Node struct {
	next  *Node
	value int
}

type LinkedList struct {
	head *Node
	tail *Node
}

func (l *LinkedList) AddInTail(item Node) {
	if l.head == nil {
		l.head = &item
	} else {
		l.tail.next = &item
	}

	l.tail = &item
}

func (l *LinkedList) Count() int {
	count := 0
	node := l.head
	for node != nil {
		count++
		node = node.next
	}
	return count
}

func (l *LinkedList) Find(n int) (Node, error) {
	node := l.head
	for node != nil {
		if node.value == n {
			return *node, nil
		}
		node = node.next
	}
	return Node{}, errors.New("node not found")
}

func (l *LinkedList) FindAll(n int) []Node {
	var nodes []Node
	curr := l.head
	for curr != nil {
		if curr.value == n {
			nodes = append(nodes, *curr)
		}
		curr = curr.next
	}
	return nodes
}

func (l *LinkedList) Delete(n int, all bool) {
	if all {
		l.deleteAll(n)
		return
	}

	l.deleteFirst(n)
}

func (l *LinkedList) Insert(after *Node, add Node) {
	if l.head == nil {
		l.head = &add
		l.tail = &add
		add.next = nil
		return
	}
	add.next = after.next
	after.next = &add
	if l.tail == after {
		l.tail = &add
	}
}

func (l *LinkedList) InsertFirst(first Node) {
	first.next = l.head
	l.head = &first
	if l.tail == nil {
		l.tail = &first
	}
}

func (l *LinkedList) Clean() {
	l.head = nil
	l.tail = nil
}

func (l *LinkedList) deleteFirst(n int) {
	if l.head == nil {
		return
	}
	if l.head.value == n {
		l.head = l.head.next
		if l.head == nil {
			l.tail = nil
		}
		return
	}
	prev := l.head
	curr := prev.next
	for curr != nil {
		if curr.value == n {
			prev.next = curr.next
			if curr == l.tail {
				l.tail = prev
			}
			return
		}
		prev = curr
		curr = curr.next
	}
}

func (l *LinkedList) deleteAll(n int) {
	for l.head != nil && l.head.value == n {
		l.head = l.head.next
	}
	if l.head == nil {
		l.tail = nil
		return
	}
	prev := l.head
	curr := prev.next
	for curr != nil {
		if curr.value == n {
			prev.next = curr.next
			if curr == l.tail {
				l.tail = prev
			}
			curr = prev.next
		} else {
			prev = curr
			curr = curr.next
		}
	}
}
